mythril/laser/ethereum/state/calldata.py

```diff
@@ -21,84 +11,133 @@ class CalldataType(Enum):
-            self._calldata = []
-            self.calldatasize = BitVecVal(len(starting_calldata), 256)
-            self.concrete = True
-        else:
-            self._calldata = Array(
-                "{}_calldata".format(self.tx_id), BitVecSort(256), BitVecSort(8)
-            )
-            self.calldatasize = BitVec("{}_calldatasize".format(self.tx_id), 256)
-            self.concrete = False
-
-        if self.concrete:
-            for calldata_byte in starting_calldata:
-                if type(calldata_byte) == int:
-                    self._calldata.append(BitVecVal(calldata_byte, 8))
-                else:
-                    self._calldata.append(calldata_byte)
-
-    def concretized(self, model):
-        result = []
-        for i in range(
-            get_concrete_int(model.eval(self.calldatasize, model_completion=True))
-        ):
-            result.append(
-                get_concrete_int(model.eval(self._calldata[i], model_completion=True))
-            )
 
+    @property
+    def calldatasize(self) -> ExprRef:
+        """
+        :return: Calldata size for this calldata object
+        """
+        result = self.size
+        if isinstance(result, int):
+            return BitVecVal(result, 256)
         return result
 
-    def get_word_at(self, index: int):
-        return self[index : index + 32]
+    def get_word_at(self, offset: int) -> ExprRef:
+        """ Gets word at offset"""
+        return self[offset : offset + 32]
 
     def __getitem__(self, item: Union[int, slice]) -> Any:
+        if isinstance(item, int) or isinstance(item, ExprRef):
+            return self._load(item)
+
         if isinstance(item, slice):
-            start, step, stop = item.start, item.step, item.stop
+            start = 0 if item.start is None else item.start
+            step = 1 if item.step is None else item.step
+            stop = self.size if item.stop is None else item.stop
+
             try:
-                if start is None:
-                    start = 0
-                if step is None:
-                    step = 1
-                if stop is None:
-                    stop = self.calldatasize
                 current_index = (
                     start if isinstance(start, BitVecRef) else BitVecVal(start, 256)
                 )
-                dataparts = []
+                parts = []
                 while simplify(current_index != stop):
-                    dataparts.append(self[current_index])
+                    parts.append(self._load(current_index))
                     current_index = simplify(current_index + step)
             except Z3Exception:
                 raise IndexError("Invalid Calldata Slice")
 
-            values, constraints = zip(*dataparts)
-            result_constraints = []
-            for c in constraints:
-                result_constraints.extend(c)
-            return simplify(Concat(values)), result_constraints
+            return simplify(Concat(parts))
```

Comment:
I think it makes more sense and is more consistent with Python to return a list when `Calldata` is accessed as a slice rather than concatenating everything together. The concatenation should instead happen in `get_word_at`.

wdyt?